### PR TITLE
add diversity metrics 07022021

### LINKED ISSUE #276: Additional TCR diversity metrics besides alpha_diversity?
Hi scirpy team,

I was wondering if more TCR diversity metrics such as inverse Simpson index could be added to scirpy in addition to alpha_diversity? Thanks very much!

### PR DESCRIPTION
**Pull request by naity**
Closes #276

### CODE REVIEW

File: scirpy/_tools/_diversity.py [hunk 1]
```diff
@@ -12,11 +13,17 @@ def alpha_diversity(
...
     *,
     target_col: str = "clone_id",
+    method: str = "normalized shannon entropy",
     inplace: bool = True,
     key_added: Union[None, str] = None,
 ) -> pd.DataFrame:
     """Computes the alpha diversity of clonotypes within a group.
 
+    Use a method out of normalized shannon entropy, D50, and `scikit-bio’s alpha diversity metrics 
+    <http://scikit-bio.org/docs/0.4.2/generated/skbio.diversity.alpha.html#module-skbio.diversity.alpha >`,
```
Comment (grst) [inline scirpy/_tools/_diversity.py:23]:
```suggestion
    <http://scikit-bio.org/docs/0.4.2/generated/skbio.diversity.alpha.html#module-skbio.diversity.alpha>`_,
```

File: scirpy/_tools/_diversity.py [hunk 2]
```diff
@@ -12,11 +13,17 @@ def alpha_diversity(
...
+    
+
+    Normalized shannon entropy:
```
Comment (grst) [inline scirpy/_tools/_diversity.py:26]: Could you please add a short description of D50 (and ideally, a reference) here?

File: scirpy/_tools/_diversity.py [hunk 3]
```diff
@@ -12,11 +13,17 @@ def alpha_diversity(
     groupby: str,
     *,
     target_col: str = "clone_id",
+    method: str = "normalized shannon entropy",
```
Comment (grst) [inline scirpy/_tools/_diversity.py:16]:
```suggestion
    method: str = "normalized_shannon_entropy",
```

File: scirpy/_tools/_diversity.py [hunk 4]
```diff
@@ -65,8 +91,24 @@ def _shannon_entropy(freq):
...
+            # calculate frequencies for these two metrics
+            tmp_freqs = tmp_counts / np.sum(tmp_counts)
+            np.testing.assert_almost_equal(np.sum(tmp_freqs), 1)
+
+            if method == "normalized shannon entropy":
+                diversity[k] = _shannon_entropy(tmp_freqs)
+            else:
+                diversity[k] = _dxx(tmp_freqs, perc=0.5)
+        
+        elif method in skbio.diversity.get_alpha_diversity_metrics():
```
Comment (grst) [inline scirpy/_tools/_diversity.py:105]: I'm a bit wary of adding too many dependencies... Could you please make skbio an *optional dependency*? See an example [here](https://github.com/icbi-lab/scirpy/blob/507c943e50a2d2ba5b0011575030db85dd5b316b/scirpy/util/graph/_component_layout.py#L87-L94).

File: scirpy/_tools/_diversity.py [hunk 5]
```diff
@@ -65,8 +91,24 @@ def _shannon_entropy(freq):
...
+            # skbio.diversity takes count vectors as input
+            diversity[k] = skbio.diversity.alpha_diversity(method, tmp_counts)
+        
+        else:
+            # raise if method is not supported
+            raise ValueError(f"{method} diversity metric is not supported")
```
Comment (grst) [inline scirpy/_tools/_diversity.py:111]: What do you think of, additionally, adding the option to pass a callback function to `method`? Like that users could bring their own diversity metrics.

File: scirpy/_tools/_diversity.py [hunk 6]
```diff
@@ -65,8 +91,24 @@ def _shannon_entropy(freq):
...
 
     if inplace:
         key_added = "alpha_diversity_" + target_col if key_added is None else key_added
```
Comment (grst) [inline scirpy/_tools/_diversity.py:114]:
```suggestion
        key_added = f"alpha_diversity_{target_col}_{method}" if key_added is None else key_added
```

File: scirpy/_plotting/_diversity.py [hunk 7]
```diff
@@ -13,6 +13,7 @@ def alpha_diversity(
     groupby: str,
     *,
     target_col: str = "clone_id",
+    method: str = "normalized shannon entropy",
```
Comment (grst) [inline scirpy/_plotting/_diversity.py:16]:
```suggestion
    method: str = "normalized_shannon_entropy",
```

File: scirpy/_tools/_diversity.py [hunk 8]
```diff
@@ -12,11 +13,17 @@ def alpha_diversity(
     groupby: str,
     *,
     target_col: str = "clone_id",
+    method: str = "normalized shannon entropy",
     inplace: bool = True,
     key_added: Union[None, str] = None,
```
Comment (grst) [inline scirpy/_tools/_diversity.py:18]: Maybe worth passing `**kwargs` to the method function? Would allow to specify other values than 50 to `dxx`.

**Review by grst — CHANGES_REQUESTED** [review]
Hi @naity, 

this looks great, thanks for adding the additional metrics! 

I added a few comments to the code. It would also be great if you could add a test for DXX and one of the scikit-bio metrics [here](https://github.com/icbi-lab/scirpy/blob/6ccaa6673d3d043e1e7988e5c92c037ad2ffc713/scirpy/tests/test_tools.py#L270-L282) to make sure everything works as expected. 

I hope I didn't overwhelm you with all the comments, I can always take over and finish it up if it's getting too much!

File: scirpy/_tools/_diversity.py (hunk 5 above)
Comment (naity) [reply scirpy/_tools/_diversity.py:111]: This is a good point. However, should we provide example code and specify whether they use freqs or counts?

**Comment by naity** [comment]
> Hi @naity,
> 
> this looks great, thanks for adding the additional metrics!
> 
> I added a few comments to the code. It would also be great if you could add a test for DXX and one of the scikit-bio metrics [here](https://github.com/icbi-lab/scirpy/blob/6ccaa6673d3d043e1e7988e5c92c037ad2ffc713/scirpy/tests/test_tools.py#L270-L282) to make sure everything works as expected.
> 
> I hope I didn't overwhelm you with all the comments, I can always take over and finish it up if it's getting too much!

Thank you for your valuable suggestions! 

I have addressed your comments and added tests. I was wondering if you would teach me how to run the tests?

**Comment by grst** [comment]
> I have addressed your comments and added tests. I was wondering if you would teach me how to run the tests?

Seems like I need to add developer docs at some point :) 

Locally, you can run them simply by invoking `pytest` on the repository root. The GitHub CI will also automatically run them on every push (I had to approve it first, since you are a first-time contributor, that's a safeguard github put into place against abusive cryptominers :roll_eyes:) 

Currently it's failing because the code is not formatted using [black](https://github.com/psf/black). The easiest way to ensure your code is always formatted correctly is to set-up black [in your IDE](https://black.readthedocs.io/en/stable/integrations/editors.html) or to use `pre-commit` (`pip install pre-commit && pre-commit intall` in the repo root).

File: scirpy/_tools/_diversity.py [hunk 9]
```diff
@@ -65,11 +100,34 @@ def _shannon_entropy(freq):
...
+                    "`scikit-bio`. You can install it with "
+                    "`pip install scikit-bio`."
+                    )
+            else:
+                # skbio.diversity takes count vectors as input and takes care of unknown metrics
+                diversity[k] = skbio.diversity.alpha_diversity(metric, tmp_counts).values[0]
 
     if inplace:
-        key_added = "alpha_diversity_" + target_col if key_added is None else key_added
+        key_added = f"alpha_diversity_{target_col}_{metric}" if key_added is None else key_added
```
Comment (grst) [inline scirpy/_tools/_diversity.py:130]: I just realized this is becoming a monster of a column name. 

I'm wondering if we should just use `f"{metric}_{target_col}"` instead...

**Comment by grst** [comment]
Don't worry about the docs failing on `ubuntu / python 3.8`. This has something to do with the deployment not working 
from PRs made from someone else's fork. The other checks should eventually pass, though!

**Comment by naity** [comment]
> > I have addressed your comments and added tests. I was wondering if you would teach me how to run the tests?
> 
> Seems like I need to add developer docs at some point :)
> 
> Locally, you can run them simply by invoking `pytest` on the repository root. The GitHub CI will also automatically run them on every push (I had to approve it first, since you are a first-time contributor, that's a safeguard github put into place against abusive cryptominers 🙄)
> 
> Currently it's failing because the code is not formatted using [black](https://github.com/psf/black). The easiest way to ensure your code is always formatted correctly is to set-up black [in your IDE](https://black.readthedocs.io/en/stable/integrations/editors.html) or to use `pre-commit` (`pip install pre-commit && pre-commit intall` in the repo root).

Great, I have reformatted the code using black. Hopefully it works this time. Thank you!

File: scirpy/_tools/_diversity.py (hunk 5 above)
Comment (grst) [reply scirpy/_tools/_diversity.py]: I think it's ok to use counts always, as its trivial to calculate frequencies from that. In that case I'd also adapt the two internal metrics to accept counts.

Comment (naity) [reply scirpy/_tools/_diversity.py]: Sounds good. I have modified `_shannon_entropy` and `_dxx` to take counts instead and added support for custom metric function. Please let me know what you think. Thank you!

Comment (naity) [reply scirpy/_tools/_diversity.py]: Hi @grst, I wanted to follow up with you. Please let me know if you have further suggestions. Thanks!

**Comment by grst** [comment]
Sorry for the delay! I pushed some updates in https://github.com/icbi-lab/scirpy/pull/277/commits/f3ad9d53b338e62c68eed0a2a62e34b6c0770247 to make the documentation a bit nicer. 

I suggest you take a final look if I screwed something up, otherwise let's merge this once the CI passes! :rocket:

**Comment by naity** [comment]
> Sorry for the delay! I pushed some updates in [f3ad9d5](https://github.com/icbi-lab/scirpy/commit/f3ad9d53b338e62c68eed0a2a62e34b6c0770247) to make the documentation a bit nicer.
> 
> I suggest you take a final look if I screwed something up, otherwise let's merge this once the CI passes! 🚀

Great, thank you for making it so much better! 

It looks like there is a small issue with the kwargs passed to the _dxx function, so I made a small change to detect percentage in kwargs. Please let me know if that works. Thanks!

**Comment by grst** [comment]
> It looks like there is a small issue with the kwargs passed to the _dxx function, so I made a small change to detect percentage in kwargs. Please let me know if that works. Thanks!

Strictly speaking this would not have been necessary, as the percentage argument would already have been passed on via the `**kwargs` argument expansion. But your version will produce a better error message, so let's keep that!

**Comment by grst** [comment]
I will try to pull in a few more bugfixes within the next 1-2 weeks and then make a release including the diversity metrics. 

Thanks again for your contribution! It's always a pleasure if someone puts a PR together :)

**Comment by naity** [comment]
> I will try to pull in a few more bugfixes within the next 1-2 weeks and then make a release including the diversity metrics.
> 
> Thanks again for your contribution! It's always a pleasure if someone puts a PR together :)

Wonderful, thank you so much for your help and suggestions! It was a great pleasure to work with you and contribute to this amazing package. I'm looking forward to collaborating with you again in the future!

**Comment by zktuong** [comment]
hi @grst @naity . The current description says `...amounting to >50% of the total sequencing reads` - looking at the implementation, am i wrong to infer that what's counted (to get the 50% threshold) is the number of `clone_id` and not sequencing reads per se?

https://github.com/scverse/scirpy/blob/443e59e6245b917e87972f87df350ae4f429d011/src/scirpy/tl/_diversity.py#L111-L118

https://github.com/scverse/scirpy/blob/443e59e6245b917e87972f87df350ae4f429d011/src/scirpy/tl/_diversity.py#L123-L124

**Comment by grst** [comment]
Thanks @zktuong, fixed in https://github.com/scverse/scirpy/pull/499. 

I guess "sequencing reads" is a leftover from the patent from which the approach was adapted.

**Pull request merged**